nexus/src/app/crucible.rs

```diff
@@ -0,0 +1,1051 @@
+                    "region_id" => %region_id,
+                    "snapshot_id" => %snapshot_id,
+                );
+
+                match e {
+                    ProgenitorOperationRetryError::Gone => {
+                        // Return Ok if the dataset's agent is gone, no delete
+                        // call is required.
+                        Ok(())
+                    }
+
+                    ProgenitorOperationRetryError::GoneCheckError(_) => {
+                        Err(Error::internal_error(
+                            "insufficient permission for crucible_agent_gone_check"
+                        ))
+                    }
+
+                    ProgenitorOperationRetryError::ProgenitorError(e) => match e {
+                        crucible_agent_client::Error::ErrorResponse(rv) => {
+                            if rv.status().is_client_error() {
+                                Err(Error::invalid_request(&rv.message))
+                            } else {
+                                Err(Error::internal_error(&rv.message))
+                            }
+                        }
+
+                        _ => {
+                            Err(Error::internal_error(
+                                "unexpected failure during `region_delete_snapshot`",
+                            ))
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Call out to a Crucible agent to delete a region
+    async fn delete_crucible_region(
+        &self,
+        log: &Logger,
+        dataset: &db::model::Dataset,
+        region_id: Uuid,
+    ) -> Result<(), Error> {
+        // If the region never existed, then a `GET` will return 404, and so
+        // will a `DELETE`. Catch this case, and return Ok if the region never
+        // existed.  This can occur if an `ensure_all_datasets_and_regions`
+        // partially fails.
+
+        match self.maybe_get_crucible_region(log, dataset, region_id).await {
+            Ok(Some(_)) => {
+                // region found, proceed with deleting
+            }
+
+            Ok(None) => {
+                // region never exited, return Ok
+                return Ok(());
+            }
+
+            // Return Ok if the dataset's agent is gone, no delete call
+            // is required.
+            Err(Error::Gone) => return Ok(()),
+
+            Err(e) => return Err(e),
+        }
+
+        // Past here, the region exists (or existed at some point): ensure it is
+        // deleted. Request the deletion (which is idempotent), then wait for
+        // the appropriate state change.
+
+        self.request_crucible_region_delete(log, dataset, region_id).await?;
+
+        // Wait until the region is deleted
+
+        backoff::retry_notify(
+            backoff::retry_policy_internal_service_aggressive(),
+            || async {
+                let region = match self.maybe_get_crucible_region(
```

Comment:
Maybe make a note that `self.maybe_get_crucible_region` will check for permanent errors and so we don't need to add a check to this loop.

Reply:
Sorry, I don't quite grok this comment - this section matches against `Err(e)`?

Reply:
Could be outdated. In either case, I wasn't even clear enough for myself to remember what I was talking about. Feel free to ignore it.